### PR TITLE
ARROW-7409: [C++][Python] Windows link error LNK1104: cannot open file 'python37_d.lib'

### PR DESCRIPTION
**Pull request by raulbocanegra**
Fix linking against Python in debug mode on Windows

### CODE REVIEW

File: cpp/cmake_modules/FindPythonLibsNew.cmake [hunk 1]
```diff
@@ -246,7 +248,7 @@ FUNCTION(PYTHON_ADD_MODULE _NAME )
       SET_TARGET_PROPERTIES(${_NAME} PROPERTIES LINK_FLAGS
                           "-undefined dynamic_lookup")
     ELSEIF(MSVC)
-      target_link_libraries(${_NAME} ${PYTHON_LIBRARIES})
+      target_link_libraries(${_NAME} ${PYTHON_LIBRARIES} $<$<CONFIG:Debug>:${PYTHON_DEBUG_LIBRARIES}> $<$<CONFIG:Release>:${PYTHON_LIBRARIES}>)
```
Comment (kou) [inline cpp/cmake_modules/FindPythonLibsNew.cmake:251]: `${PYTHON_LIBRARIES}` is duplicated in the `Release` configuration.

File: cpp/src/arrow/python/CMakeLists.txt [hunk 2]
```diff
@@ -60,7 +60,11 @@ endif()
 
 set(ARROW_PYTHON_SHARED_LINK_LIBS arrow_shared)
 if(WIN32)
-  list(APPEND ARROW_PYTHON_SHARED_LINK_LIBS ${PYTHON_LIBRARIES} ${PYTHON_OTHER_LIBS})
+  if(ARROW_USE_DEPRECATED_FIND_PYTHONLIBS)
+    list(APPEND ARROW_PYTHON_SHARED_LINK_LIBS $<$<CONFIG:Debug>:${PYTHON_DEBUG_LIBRARIES}> $<$<CONFIG:Release>:${PYTHON_LIBRARIES}>)
+  else()
+    list(APPEND ARROW_PYTHON_SHARED_LINK_LIBS ${PYTHON_OTHER_LIBS})
```
Comment (kou) [inline cpp/src/arrow/python/CMakeLists.txt:66]: `${PYTHON_OTHER_LIBS}` should be always used.

**Review by kou — COMMENTED** [review]
BTW, does the following change fix this problem?

```diff
diff --git a/cpp/cmake_modules/FindPythonLibsNew.cmake b/cpp/cmake_modules/FindPythonLibsNew.cmake
index 581bba9d4..a5b7656a0 100644
--- a/cpp/cmake_modules/FindPythonLibsNew.cmake
+++ b/cpp/cmake_modules/FindPythonLibsNew.cmake
@@ -149,8 +149,13 @@ string(REGEX REPLACE "\\\\" "/" PYTHON_SITE_PACKAGES ${PYTHON_SITE_PACKAGES})
 if(CMAKE_HOST_WIN32)
   # Appease CMP0054
   if(CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
-    set(PYTHON_LIBRARY
-      "${PYTHON_PREFIX}/libs/Python${PYTHON_LIBRARY_SUFFIX}.lib")
+    if(PYTHON_IS_DEBUG)
+      set(PYTHON_LIBRARY
+        "${PYTHON_PREFIX}/libs/Python${PYTHON_LIBRARY_SUFFIX}.lib")
+    else()
+      set(PYTHON_LIBRARY
+        "${PYTHON_PREFIX}/libs/Python${PYTHON_LIBRARY_SUFFIX}_d.lib")
+    endif()
   else()
     find_library(PYTHON_LIBRARY
         NAMES "python${PYTHON_LIBRARY_SUFFIX}"
```

File: cpp/cmake_modules/DefineOptions.cmake [hunk 3]
```diff
@@ -120,6 +120,8 @@ if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
 
   define_option(ARROW_GGDB_DEBUG "Pass -ggdb flag to debug builds" ON)
 
+  define_option(ARROW_USE_DEPRECATED_FIND_PYTHONLIBS "Force using deprecated FindPythonLibsNew script" OFF)
```
[the remark this answers is not in the document]
Comment (raulbocanegra) [reply cpp/cmake_modules/DefineOptions.cmake:123]: 👍

File: cpp/cmake_modules/FindPythonLibsNew.cmake (hunk 1 above)
Comment (raulbocanegra) [reply cpp/cmake_modules/FindPythonLibsNew.cmake:251]: :+1:

File: cpp/src/arrow/python/CMakeLists.txt [hunk 4]
```diff
@@ -87,6 +94,10 @@ add_arrow_lib(arrow_python
               EXTRA_INCLUDES
               "${ARROW_PYTHON_INCLUDES}")
 
+if(WIN32 AND TARGET arrow_python_shared AND NOT ARROW_USE_DEPRECATED_FIND_PYTHONLIBS)
```
Comment (pitrou) [inline cpp/src/arrow/python/CMakeLists.txt:97]: Why is this complication needed?

File: cpp/src/arrow/python/CMakeLists.txt [hunk 5]
```diff
@@ -60,7 +60,14 @@ endif()
 
 set(ARROW_PYTHON_SHARED_LINK_LIBS arrow_shared)
 if(WIN32)
-  list(APPEND ARROW_PYTHON_SHARED_LINK_LIBS ${PYTHON_LIBRARIES} ${PYTHON_OTHER_LIBS})
+  if(ARROW_USE_DEPRECATED_FIND_PYTHONLIBS)
+    list(APPEND ARROW_PYTHON_SHARED_LINK_LIBS 
+         $<$<CONFIG:Debug>:${PYTHON_DEBUG_LIBRARIES}> 
+         $<$<CONFIG:Release>:${PYTHON_LIBRARIES}>
```
Comment (pitrou) [inline cpp/src/arrow/python/CMakeLists.txt:66]: Can't you factor out the debug/release library switch?

**Review by pitrou — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/python/CMakeLists.txt (hunk 4 above)
Comment (raulbocanegra) [reply cpp/src/arrow/python/CMakeLists.txt:97]: Well, the point here is that `PYTHON_LIBRARIES` was originally appended to `ARROW_PYTHON_SHARED_LINK_LIBS` as you can see on line 63 (red) above.
Because of that, we ended in a situation in which we have two different kinds of items in the `ARROW_PYTHON_SHARED_LINK_LIBS` list, the first item is the target `arrow_shared` and a second item in the form of `debug`, `optimized` keyword followed by the library path. It is described at the first and last items in the list [here](https://cmake.org/cmake/help/latest/command/target_link_libraries.html#id2). 
I don't know if mixing different types of items in a list variable is supported, but the fact is that it fails, so I have to add that dependency on a second step here, after `add_arrow_lib`. We only need to do this on Windows, when we are not using the deprecated scripts, and of course, if the target is defined. 
I know it is not the prettiest if branch you can find, but I couldn't find another way. 
If the description of the problem above helps you finding a better solution I am happy to change it.

File: cpp/src/arrow/python/CMakeLists.txt (hunk 5 above)
Comment (raulbocanegra) [reply cpp/src/arrow/python/CMakeLists.txt:66]: Hi, `PYTHON_DEBUG_LIBRARIES` is set on the deprecated `FindPythonLibsNew` script. If I don't use it here, it won't link against `python_d` and the linking stage will fail.

**Comment by kou** [comment]
@raulbocanegra Could you confirm that https://github.com/apache/arrow/pull/7284#pullrequestreview-419603540 without your changes can fix this problem?

**Comment by raulbocanegra** [comment]
> 
> 
> @raulbocanegra Could you confirm that [#7284 (review)](https://github.com/apache/arrow/pull/7284#pullrequestreview-419603540) without your changes can fix this problem?

Sorry, that change does not work because `PYTHON_IS_DEBUG` is not set to 1 unless `PYTHON_EXECUTABLE` is `python_d`, but `find_package(PythonInterp)` does not return `python_d`, but `python` release executable.

**Comment by kou** [comment]
Thanks for confirming it.

Could you confirm one more thing?
Does https://github.com/apache/arrow/pull/7284#pullrequestreview-419603540 without your changes and `cmake ... -DPYTHON_EXECUTABLE=python_d` fix this problem?

**Comment by wesm** [comment]
@raulbocanegra ping

**Comment by raulbocanegra** [comment]
Hi!
First of all, sorry for the delay guys. I tried the `-DPYTHON_EXECUTABLE=python_d` but it then raise an error related to some numpy library.
I am really busy this week. Please, give me a few days more to give you some more information.

**Comment by kou** [comment]
No problem. Thanks for confirming.

**Comment by raulbocanegra** [comment]
Hi again,
The error related to numpy seems to be that I have a release version of numpy, but I am building a debug version of Arrow.
[Numpy docs](https://numpy.org/doc/stable/user/troubleshooting-importerror.html#debug-build-on-windows) recommends building using `RelWithDebInfo` or recompile the entire stack of python modules including numpy in Debug mode which it's a real pain in case you have a lot of modules.
To sum up, I will try to stick with `RelWithDebInfo` build type for Windows, so I feel that this PR could be closed.

**Comment by kou** [comment]
OK. I close this.
